### PR TITLE
Allow for different versions of a question title depending on a conditional question

### PR DESCRIPTION
**Pull request by shahi645**
### What is the context of this PR?
We have a use case in SR where the user answers a proxy question, e.g. "Who are you answering on behalf of?". This means that we want subsequent questions to be asked differently depending on the users answers. Previously this meant using routing rules and having a lot of excess schema.
This PR introduces a `titles` object to question, this can be used instead of a single `title` value. The titles object consists of a `default value` with is a single title, with optional extra `titles` that are accompanied with a `when` rule (`when` functionality already exists in routing rules, so we have re-used that code). When the question title is rendered each `when clause` is evaluated and if true returns its given `title`. If no `when clauses` evaluate true then the `default title` is returned and rendered.

**UPDATE**
We have added changes that now mean all getting of question titles are handled by our templating utils and template logic on selecting the question title is handled in our jinja filters.

***NOTE***
Currently we have a bug with repeating groups that are as follows: 
- The summary only renders the latest group instance of a repeating group. (existing bug)

- When evaluating a `when` clause in a repeating group greater than `group_instance` **0** it defaults to 0 `group_instance` for summary rendering, as summary has no knowledge of which group instance the question/answer has come from.

### How to review 
Run our test_schemas created and see if the changes are correct. 
Also does the code changes make sense as to where they are.

### CODE REVIEW

**Comment by shahi645** [comment]
please note travis build wont pass until https://github.com/ONSdigital/eq-survey-runner/pull/1605 is merged

File: app/questionnaire/questionnaire_schema.py
Comment (nealedj) [inline app/questionnaire/questionnaire_schema.py]: There's already a `get_answers_by_id_for_block` method. Could we follow the convention and rename this to `get_answer_ids_for_question`?

Following the convention you'd have

```
    def get_answers_by_id_for_question(self, question_id):
        question = self.get_question(question_id)
        if question:
            return {
                answer['id']: answer
                for answer in question['answers']
            }

        return {}

    def get_answer_ids_for_question(self, question_id):
        return list(self.get_answers_by_id_for_question(question_id).keys())

    def get_answers_for_question(self, question_id):
        return list(self.get_answers_by_id_for_question(question_id).values())
```

File: app/questionnaire/questionnaire_schema.py
Comment (nealedj) [inline app/questionnaire/questionnaire_schema.py]: Lots of nesting here. Making it quite tricky to read.

Possible refactor:

```
        for title in question.get('titles', []):
            if 'when' not in title:
                continue

        when_clauses = title.get('when')

        for when_clause in when_clauses:
            for answer_id in answer_ids:
                when_id = when_clause.get('id')
                if answer_id not in dependencies[when_id]:
                    dependencies[when_id].append(answer_id)
```

File: app/questionnaire/questionnaire_schema.py
Comment (nealedj) [inline app/questionnaire/questionnaire_schema.py]: Some of these methods are called `_set_xxx` and others `_get_xxx` but it's not clear why. As they're all mutating the `dependencies` object `set` might be more appropriate.

File: app/questionnaire/questionnaire_schema.py
Comment (nealedj) [inline app/questionnaire/questionnaire_schema.py]: Consider making this a `defaultdict(set)` so that you can add to it without the need to check if a value is already present.

E.g.
```
>>> question_id, answer_id = 'foo', 'bar'
>>> dependencies = defaultdict(set)
>>> dependencies[question_id].add(answer_id)
>>> dependencies[question_id].add(answer_id)
{'foo': {'bar'}}
```

File: app/questionnaire/rules.py
Comment (nealedj) [inline app/questionnaire/rules.py]: Could be reduced to:
`instance  = when_specified_group_instance or group_instance`

File: app/templating/summary/question.py
Comment (nealedj) [inline app/templating/summary/question.py]: Nitpick - could be a generator:

```
when_rules = ((title['when'], title['value']) for title in titles if 'when' in title)
```

File: app/templating/summary/question.py
Comment (nealedj) [inline app/templating/summary/question.py]: Indexing a tuple rather than unpacking to variables.

Possibly more readable:

```
for when, value in when_rules:
    if evaluate_when_rules(when, metadata, answer_store, group_instance):
        return value
```

File: app/templating/summary/question.py
Comment (nealedj) [inline app/templating/summary/question.py]: What's the expected behaviour if the `title` is defined but == `""`?

Given that the templates are using the `is defined` jinja operator I'd say that for consistency we should treat empty strings as a defined title.

So this could become `if 'title' in question_schema:`

File: app/templating/summary/question.py
Comment (nealedj) [inline app/templating/summary/question.py]: Suggest that this is opened onto separate lines for readability:

```
if 'titles' in question_schema:
    return cls.get_title_from_titles(metadata, answer_store, titles, 0)
else:
    return question_schema['answers'][0]['label']
```

Comment (nealedj) [inline app/templating/summary/question.py]: We're explicitly setting the `group_instance` to `0` here but without explanation.

We know it's because summaries don't yet support repeating groups - could we put a comment to explain?

File: data/en/test_different_question_titles.json
Comment (nealedj) [inline data/en/test_different_question_titles.json]: effect ;)

File: scripts/test_schemas.sh
Comment (nealedj) [inline scripts/test_schemas.sh]: Note to remove this before merge.

File: tests/app/forms/test_questionnaire_form.py
Comment (nealedj) [inline tests/app/forms/test_questionnaire_form.py]: I don't understand what this is testing.

File: app/templating/view_context.py
Comment (nealedj) [inline app/templating/view_context.py]: I don't think the `question_titles` should be nested inside the `form` object. They don't belong to it.

In the absence of a proper view model I'd suggest patching the title directly onto the `question` objects within each `rendered_block`.

```
for question in rendered_block.get('questions'):
    question['question_title`] = get_the_title()
```

Then the templates wouldn't need any logic for working out where to get the title from.

File: app/templating/view_context.py
Comment (nealedj) [inline app/templating/view_context.py]: This is being imported so that the `get_title_from_titles` function can be called. I think that implies that the function is in the wrong place because it isn't specific to the summary.

Consider adding a `utils` module to the `templating` package for now and add the function there instead.

**Review by nealedj — CHANGES_REQUESTED** [review]
Well done - loads of good stuff here. Lots of little comments but nothing fundamental. The main thing that I think could be improved is where the titles are rendered into the view context.

General point - could the test schemas be named such that they're grouped together? Might be useful for these bigger features that require multiple schemas.

E.g. test_titles.json, test_titles_within_repeating_blocks.json, test_titles_radio_checkbox

File: app/questionnaire/questionnaire_schema.py
Comment (AndrewTorrance) [reply app/questionnaire/questionnaire_schema.py]: Done

File: app/questionnaire/questionnaire_schema.py
Comment (AndrewTorrance) [reply app/questionnaire/questionnaire_schema.py]: Did a different implementation using list comprehensions.

File: app/questionnaire/questionnaire_schema.py
Comment (AndrewTorrance) [reply app/questionnaire/questionnaire_schema.py]: That one does not directly set the self.dependencies , it returns dependencies which other code assigns to the self.dependencies so I think its  valid as it does not off itself affect the object state.

Comment (nealedj) [reply app/questionnaire/questionnaire_schema.py]: The two `_get` methods and the `_set` method both mutate the `dependencies` argument that gets passed to them - I'm not why they employ different verbs.

Generally (subjectively) I think I'd expect a method start with `get` to not mutate any state. A `set` method might mutate the state of the object.

In this case perhaps `build` might be a better word to use?

File: app/questionnaire/questionnaire_schema.py
Comment (AndrewTorrance) [reply app/questionnaire/questionnaire_schema.py]: I like the idea , but the dependencies are used by existing code , so I'd rather leave as is for now.

File: app/questionnaire/rules.py
Comment (AndrewTorrance) [reply app/questionnaire/rules.py]: good catch

File: app/templating/summary/question.py
Comment (AndrewTorrance) [reply app/templating/summary/question.py]: Changed , I was toying with a named tuple , but no need for 1 use simple unpacking is perfectly good

File: app/templating/summary/question.py
Comment (AndrewTorrance) [reply app/templating/summary/question.py]: Its is nitpicky , but valid , changed

File: app/templating/summary/question.py
Comment (AndrewTorrance) [reply app/templating/summary/question.py]: Its not that straightforward . The pre existing logic was if title =='' then use the value defined in the answer label as the title. So we changed the logic to
if schema[title] return schema[title]
if valid title in titles return that title
else return answer label for question 0 

I think should be revisited when looking at repeating groups .

File: app/templating/summary/question.py
Comment (AndrewTorrance) [reply app/templating/summary/question.py]: mmm com si com sa , changed it without the else

Comment (AndrewTorrance) [reply app/templating/summary/question.py]: Added

File: data/en/test_different_question_titles.json
Comment (AndrewTorrance) [reply data/en/test_different_question_titles.json]: :-)

File: scripts/test_schemas.sh
Comment (AndrewTorrance) [reply scripts/test_schemas.sh]: Added comment to script. PR also has do not merge on it

File: app/templating/view_context.py
Comment (AndrewTorrance) [reply app/templating/view_context.py]: We raised something similar as a possibility at the beginning but we shied away from that after sharing the ideas because of the need to know if the question was inside a repeating group / summary . We can do it but it would mean a significant rewrite .

File: app/templating/view_context.py
Comment (AndrewTorrance) [reply app/templating/view_context.py]: Your spot on. Adding it to view context caused circular dependencies . Moving out to a utils file

**Comment by shahi645** [comment]
@nealedj ready for re-review

File: app/forms/questionnaire_form.py
Comment (davecharles) [inline app/forms/questionnaire_form.py]: Bug in waiting... `group_instacne`

File: app/jinja_filters.py
Comment (davecharles) [inline app/jinja_filters.py:233]: Existing filters have adopted the convention of `format_` - could we call theses `format_question_title` etc?

File: app/jinja_filters.py
Comment (davecharles) [inline app/jinja_filters.py]: Maybe `question_title_processor` for consistency?

File: app/templating/utils.py
Comment (davecharles) [inline app/templating/utils.py:1]: Don't understand why the separate utils module.  It's very 'titles' centric so not really utils. Less cognitive load for the reader if it's just in `view_context.py` - unless there's a reason I'm missing (e.g. avoiding circ import).

File: app/views/questionnaire.py
Comment (davecharles) [inline app/views/questionnaire.py]: I think this is a bit more readable:
```python
if first_question.get('title') is not None:
    return first_question['title']
return context['question_titles'][first_question['id']]
```
Do you explicitly want to test for None and not empty str as well?

File: tests/app/forms/test_household_composition.py
Comment (davecharles) [inline tests/app/forms/test_household_composition.py:23]: The function def only has positional args so technically you don't need to name `metadata` and `group_instance` keywords. Not a big issue but I'm wondering if it says something about `generate_household_composition_form` args - do we need a `*` arg list delimiter for example?

**Review by davecharles — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py
Comment (shahi645) [reply app/jinja_filters.py:233]: i think in this instance its better as a get, we're not mutating anything here we're just getting a value based on the context

File: app/templating/utils.py
Comment (shahi645) [reply app/templating/utils.py:1]: its called in several places, so we agreed in the utils would be best served

File: tests/app/forms/test_household_composition.py
Comment (shahi645) [reply tests/app/forms/test_household_composition.py:23]: i just named them for clarity, i guess thats a better standard but not sure myself. Raises food for thought though, all tutorials i have watched have mentioned to name args regardless

File: app/jinja_filters.py
Comment (shahi645) [reply app/jinja_filters.py]: yeah can do, the others have check but you're right processor probably sounds better

File: app/views/questionnaire.py
Comment (shahi645) [reply app/views/questionnaire.py]: its tested in `test_questionnaire`

File: app/questionnaire/questionnaire_schema.py
Comment (AndrewTorrance) [reply app/questionnaire/questionnaire_schema.py]: I can see why its confusing . They are internal functions , none of which change the dependencies as stored on self. I have changed them all to gets as they get the dependencies . The actual set is done in the parse

File: tests/functional/spec/features/conditional_question_title/conditional_titles_in_repeating_groups.spec.js
Comment (davecharles) [inline tests/functional/spec/features/conditional_question_title/conditional_titles_in_repeating_groups.spec.js]: Note to remove this before merge

File: tests/functional/spec/features/conditional_question_title/conditional_titles_in_repeating_groups.spec.js
Comment (davecharles) [inline tests/functional/spec/features/conditional_question_title/conditional_titles_in_repeating_groups.spec.js]: Need to remove this @watch

Comment (AndrewTorrance) [reply tests/functional/spec/features/conditional_question_title/conditional_titles_in_repeating_groups.spec.js]: Done

**Review by davecharles — APPROVED** [review]

File: tests/app/forms/test_household_composition.py
Comment (nealedj) [reply tests/app/forms/test_household_composition.py:23]: As a rule I never name positional args and always name kwargs. TBH I've never questioned it but all python devs I've worked with have done the same.

File: app/templating/utils.py
Comment (nealedj) [reply app/templating/utils.py:1]: Ideally this logic would form part of a `Question` view model of some type. But we don't have that concept and it's quite a big one to introduce.

Separate module is because it's used by both the context builder module and the summary view models.

And yep, there would have been a circular import if it was defined in `view_context.py` since that imports the summary view model which depends on this function.

File: app/views/questionnaire.py
Comment (nealedj) [inline app/views/questionnaire.py]: This might make more sense as `if 'title' in first_question`. My reasoning being that as a rule if `title` is defined on a schema then it should take precedence over `titles` - even if `title` is an empty string.

All this check does is adds implicit behaviour if somebody were to define `"title": null` (which validator wouldn't allow anyway).

**Review by nealedj — APPROVED** [review]
Really good work chaps. I've left one minor comment to consider but I'm not blocking it based on that.

**Comment by dwyeradam** [comment]
Apparently I'm the Simon Cowell of eQ reviews and the designated merger here. Just doing a pass of the code and i'll merge it when i'm done.

File: app/questionnaire/questionnaire_schema.py
Comment (dwyeradam) [inline app/questionnaire/questionnaire_schema.py]: This is a type of question level dependency so rather than loop through the questions twice it's worth calling `_get_titles_dependencies_for_question(...)` within `_get_question_level_dependencies()`. 
It may also be worth moving the existing code in `_get_question_level_dependencies()` into it's own method, something like ``_get_calculations_dependencies_for_question(...)``

File: app/questionnaire/questionnaire_schema.py
Comment (dwyeradam) [inline app/questionnaire/questionnaire_schema.py]: I'm not a fan of mutating parameters. We'd have to test if we can safely merge the dictionary though if keys already exist. We may already have an issue here.

**Review by dwyeradam — CHANGES_REQUESTED** [review]

File: app/questionnaire/dependencies.py
Comment (nealedj) [inline app/questionnaire/dependencies.py]: Instead of holding a defaultdict internally maybe consider extending a `defaultdict` for this class.

```
class Dependencies(defaultdict):
  def __init__(self):
    super().__init__(set)

  def update(*args, **kwargs):    
    for other in args:  
      self._merge(other)
    if kwargs:
      self._merge(kwargs)

  def _merge(self, other):
    for key, dependencies in other.items():
      self[key] |= dependencies
```

(untested)

File: app/questionnaire/dependencies.py
Comment (nealedj) [inline app/questionnaire/dependencies.py]: I'm not sure it's a good idea to implement the `+` operator for the `Dependencies` object when the `Dependencies` object itself is largely acting as a dict. Dicts can't be added together, that's what `update()` is for, so I'd suggest implementing that instead.

File: app/questionnaire/dependency_builder.py
Comment (nealedj) [inline app/questionnaire/dependency_builder.py]: Since the dependencies state is held and managed by the `Dependencies` type I don't think there's much value in this being a class itself.

In a static language this would be a static class - it holds no state. In python I'd suggest that it's more pythonic to define the methods as functions on the module instead.

**Review by nealedj — CHANGES_REQUESTED** [review]

**Review by dwyeradam — APPROVED** [review]

**Review by nealedj — APPROVED** [review]

**Pull request merged**
